src/Passes/Source/AllocationManager/IAllocationManager.hpp

```diff
@@ -0,0 +1,74 @@
+#pragma once
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Llvm/Llvm.hpp"
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace microsoft
+{
+namespace quantum
+{
+    /// Interface class for allocation management. This interface provides means to allocate and release
+    /// statically allocated resources such as qubits and results. In a future version, it may be
+    /// extended with get and store in order to support Arrays and Tuples.
+    class IAllocationManager
+    {
+      public:
+        using Address              = uint64_t;    ///< Value type for address
+        using Index                = uint64_t;    ///< Index type used to access an array element.
+        using String               = std::string; ///< String type used for naming segments.
+        using AllocationManagerPtr = std::shared_ptr<IAllocationManager>; ///< Pointer interface.
+
+        // Construction, moves and copies
+        //
+        IAllocationManager(IAllocationManager const&) = delete;
+        IAllocationManager(IAllocationManager&&)      = delete;
+        IAllocationManager& operator=(IAllocationManager const&) = delete;
+        IAllocationManager& operator=(IAllocationManager&&) = delete;
+
+        virtual ~IAllocationManager();
+
+        // Interface
+        //
+
+        /// Abstract member function to allocate an element or sequence of elements. The developer
+        /// should not assume continuity of the address segment as this is not guaranteed. Note this
+        /// function may throw if allocation is not possible.
+        virtual Address allocate(String const& name = "", Index const& size = 1) = 0;
+
+        /// Abstract member function to release a previously allocated function. Note this function may
+        /// throw if an invalid address is passed.
+        virtual void release(Address const& address) = 0;
+
+        /// Abstract member function to reset the allocation manager. This function clears all allocations
+        /// and resets all statistics.
+        virtual void reset() = 0;
+
+        // Statistics
+        //
+
+        /// Current number of registers in use. This function is used to inquire about the current number
+        /// registers/resources in use.
+        uint64_t registersInUse() const;
+
+        /// Maximum number of registers in use at any one time. The maximum number of registers used at
+        /// any one time. As an example of usage, this function is useful to calculate the total number of
+        /// qubits required to execute the entry function.
+        uint64_t maxRegistersUsed() const;
```

Comment:
FYI:
From the documentation and from this comment I got an impression that 
the _qubit_ (in quantum computing) is treated as a synonym of the _register_ (in classical computing),
whereas my perception is that qubit is a synonym of _bit_ (rather than register). 
(qubit = quantum _bit_).
Based on that a better name for `maxRegistersUsed()` could be `maxBitsUsed()`.

Reply:
Yes, it is true that in many ways a qubit is closer to a bit than it is to an integer. However, you can have 1-bit registers in classical computing: It is not very common for normal CPUs but is used in FPGAs for instance. The reason why I've named it a register is that it stores a state in a similar way to a classical register. Also, note that the allocation manager is not exclusive to qubits - it also covers results and potentially could cover other things in the future.

Reply:
Having such multi-purpose allocations, I would suggest to consider refraining from using the word "Register". Something like max**Allocations**Used() could be a better match for the concept.
Up to you.

Reply:
That's a good suggestion. I will revise accordingly.